### PR TITLE
Improve exception for `validate_key`

### LINKED ISSUE #10764: Exception from `validate_key` is misleading for composite keys
When validating a key tuple with an invalid component, the exception is currently not helpful for identifying the problem and might even be misleading:

```python
import numpy as np
from dask.core import validate_key
validate_key((1, np.int64(2), 3))
```
creates the following traceback:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/hendrikmakait/projects/dask/dask/dask/core.py", line 223, in validate_key
    raise TypeError(f"Unexpected key type {type(key)} (value: {key!r})")
TypeError: Unexpected key type <class 'tuple'> (value: (1, 2, 3))
```

That is, the exception suggests that `tuple` is an invalid type a key even though it is the _only_ valid type for creating composite keys. The printed key also doesn't give any hint that the `2` is to blame here.

### PR DESCRIPTION
**Pull request by hendrikmakait**
- [x] Closes #10764
- [ ] Tests added / passed
- [x] Passes `pre-commit run --all-files`

The exception now indicates the index of the invalid key type and the traceback reveals the actual problem. The latter is important for nested tuples.

```
Traceback (most recent call last):
  File "/Users/hendrikmakait/projects/dask/dask/dask/core.py", line 235, in validate_key
    validate_key(part)
  File "/Users/hendrikmakait/projects/dask/dask/dask/core.py", line 240, in validate_key
    raise TypeError(f"Unexpected key type {typ} ({key=!r})")
TypeError: Unexpected key type <class 'numpy.int64'> (key=2)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/hendrikmakait/projects/dask/dask/dask/core.py", line 237, in validate_key
    raise TypeError(
TypeError: Composite key contains unexpected key type at _index=1 (key=(1, 2, 3))
```

### CODE REVIEW

**Review by phofl — APPROVED** [review]
lgtm

File: dask/core.py [hunk 1]
```diff
@@ -218,9 +218,26 @@ def iskey(key: object) -> bool:
...
+    typ = type(key)
+
+    if typ is tuple:
+        _index = None
+        try:
+            for _index, part in enumerate(cast(tuple, key)):
+                validate_key(part)
+        except TypeError as e:
+            raise TypeError(
+                f"Composite key contains unexpected key type at {_index=} ({key=!r})"
```
Comment (phofl) [inline dask/core.py:238]:
```suggestion
                f"Composite key contains unexpected key type at index={_index} ({key=!r})"
```

That would get ci passing

Comment (hendrikmakait) [reply dask/core.py]: I shouldn't have followed `flake8`s advice there. Reverted to `index` and ignoring the check.

Comment (phofl) [reply dask/core.py]: Sounds good to me

**Pull request merged**